File: app/src/main/java/com/nollpointer/dates/adapters/SortCardsAdapter.java
package com.nollpointer.dates.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.nollpointer.dates.Date;
import com.nollpointer.dates.R;

import java.util.ArrayList;
import java.util.List;

public class SortCardsAdapter extends RecyclerView.Adapter<SortCardsAdapter.ViewHolder> {
    private SortCardsAdapter.Listener listener;

    private int itemCount = 3;
    private List<Date> dates;

    private ArrayList<Integer> sequence = new ArrayList<>();

    public interface Listener {
        void onClick(int position);
    }

    @Override
    public SortCardsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_sort, parent, false);
        return new SortCardsAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(SortCardsAdapter.ViewHolder holder, final int position) {
        View view = holder.itemView;
        TextView mainTextView = view.findViewById(R.id.textMain);
        TextView numberTextView = view.findViewById(R.id.textNumber);
        numberTextView.setText(Integer.toString(position + 1));
        mainTextView.setText(dates.get(position).getEvent());
        sequence.add(position);
    }

    @Override
    public int getItemCount() {
        return itemCount;
    }

    public boolean onItemMove(int fromPosition, int toPosition) {
        notifyItemMoved(fromPosition, toPosition);

        int fromNumber = sequence.get(fromPosition);
        int toNumber = sequence.get(toPosition);

        sequence.set(toPosition,fromNumber);
        sequence.set(fromPosition,toNumber);

        return true;
    }


    public void setListener(SortCardsAdapter.Listener listener) {
        this.listener = listener;
    }

    public void setItemCount(int count) {
        this.itemCount = count;
    }

    public void setDates(List<Date> dates) {
        this.dates = dates;
        sequence.clear();
    }

    public List<Integer> getAnswerSequence(){
        return sequence;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ViewHolder(View view) {
            super(view);
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onClick(getAdapterPosition());
                }
            });
        }
    }
}

File: app/src/main/java/com/nollpointer/dates/sort/SortCardsAdapter.kt
package com.nollpointer.dates.sort

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.nollpointer.dates.R
import com.nollpointer.dates.other.Date
import java.util.*

class SortCardsAdapter : RecyclerView.Adapter<SortCardsAdapter.ViewHolder>() {
    private lateinit var listener: Listener
    private var itemCount = 3
    private lateinit var dates: List<Date>
    private val sequence = ArrayList<Int>()

    interface Listener {
        fun onClick(position: Int)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.card_sort, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val view = holder.itemView
        val mainTextView = view.findViewById<TextView>(R.id.sortTextMain)
        val numberTextView = view.findViewById<TextView>(R.id.sortTextNumber)
        val imageView = view.findViewById<ImageView>(R.id.sortImage)
        numberTextView.text = (position + 1).toString()
        mainTextView.text = dates[position].event
        sequence.add(position)
        imageView.visibility = View.INVISIBLE
    }

    override fun getItemCount(): Int {
        return itemCount
    }

    fun onItemMove(fromPosition: Int, toPosition: Int): Boolean {
        notifyItemMoved(fromPosition, toPosition)
        val fromNumber = sequence[fromPosition]
        val toNumber = sequence[toPosition]
        sequence[toPosition] = fromNumber
        sequence[fromPosition] = toNumber
        return true
    }

    fun setListener(listener: Listener) {
        this.listener = listener
    }

    fun setItemCount(count: Int) {
        itemCount = count
    }

    fun setDates(dates: List<Date>) {
        this.dates = dates
        sequence.clear()
    }

    val answerSequence: List<Int>
        get() = sequence

    inner class ViewHolder internal constructor(view: View?) : RecyclerView.ViewHolder(view!!)
}